File: code/zato-web-admin/src/zato/admin/web/views/security/vault/connection.py
# -*- coding: utf-8 -*-

"""
Copyright (C) 2019, Zato Source s.r.o. https://zato.io

Licensed under AGPLv3, see LICENSE.txt for terms and conditions.
"""

from __future__ import absolute_import, division, print_function, unicode_literals

# stdlib
import logging

# Zato
from zato.admin.web.forms.security.vault.connection import CreateForm, EditForm
from zato.admin.web.views import CreateEdit, Delete as _Delete, Index as _Index
from zato.common.odb.model import VaultConnection

logger = logging.getLogger(__name__)

class Index(_Index):
    method_allowed = 'GET'
    output_class = VaultConnection
    url_name = 'security-vault-conn'
    template = 'zato/security/vault/connection.html'
    service_name = 'zato.security.vault.connection.get-list'
    paginate = True

    class SimpleIO(_Index.SimpleIO):
        input_required = ('cluster_id',)
        output_required = ('id', 'name', 'url',  'default_auth_method',  'timeout',  'allow_redirects',  'tls_verify')
        output_optional = ('token', 'service_id', 'service_name')
        output_repeated = True

    def on_before_append_item(self, item):
        item.token = item.token or ''
        item.default_auth_method = item.default_auth_method or ''
        return item

    def handle(self):

        return {
            'create_form': CreateForm(self.req),
            'edit_form': EditForm(self.req, prefix='edit'),
        }

class _CreateEdit(CreateEdit):
    method_allowed = 'POST'

    class SimpleIO(CreateEdit.SimpleIO):
        input_required = ('id', 'name', 'is_active', 'url',  'default_auth_method',  'timeout',  'allow_redirects', 'sec_type',
            'tls_verify', 'cluster_id')
        input_optional = ('token', 'service_id')
        output_required = ('id', 'service_name')

    def post_process_return_data(self, return_data):
        if self.input.service_id:
            response = self.req.zato.client.invoke('zato.service.get-by-id', {
                'id': self.input.service_id,
                'cluster_id': self.req.zato.cluster_id
            })

            return_data['service_id'] = response.data.id
            return_data['service_name'] = response.data.name

        return return_data

    def success_message(self, item):
        return 'Successfully {} Vault connection `{}`'.format(self.verb, item.name)

class Create(_CreateEdit):
    url_name = 'security-vault-conn-create'
    service_name = 'zato.security.vault.connection.create'

class Edit(_CreateEdit):
    url_name = 'security-vault-conn-edit'
    form_prefix = 'edit-'
    service_name = 'zato.security.vault.connection.edit'

class Delete(_Delete):
    url_name = 'security-vault-conn-delete'
    error_message = 'Could not delete the role permission'
    service_name = 'zato.security.vault.connection.delete'

File: code/zato-web-admin/src/zato/admin/static/js/security/vault/connection.js

// /////////////////////////////////////////////////////////////////////////////

$.fn.zato.data_table.VaultConnection = new Class({
    toString: function() {
        var s = '<VaultConnection id:{0} name:{1}>';
        return String.format(s, this.id ? this.id : '(none)',
                this.name ? this.name : '(none)');
    }

});

// /////////////////////////////////////////////////////////////////////////////

$(document).ready(function() {
    $('#data-table').tablesorter();
    $.fn.zato.data_table.password_required = false;
    $.fn.zato.data_table.class_ = $.fn.zato.data_table.VaultConnection;
    $.fn.zato.data_table.new_row_func = $.fn.zato.security.vault.connection.data_table.new_row;
    $.fn.zato.data_table.parse();
    $.fn.zato.data_table.setup_forms(['name', 'url', 'timeout']);
})


$.fn.zato.security.vault.connection.create = function() {
    $.fn.zato.data_table._create_edit('create', 'Create a new Vault connection', null);
}

$.fn.zato.security.vault.connection.edit = function(id) {
    $.fn.zato.data_table._create_edit('edit', 'Update the Vault connection', id);
}

$.fn.zato.security.vault.connection.data_table.new_row = function(item, data, include_tr) {
    var row = '';

    if(include_tr) {
        row += String.format("<tr id='tr_{0}' class='updated'>", item.id);
    }

    if(data.service_name) {
        var service_name = String.format(
            '<a href="/zato/service/overview/{0}/?cluster={1}">{0}</a>', data.service_name, item.cluster_id);
    }
    else {
        var service_name = "<span class='form_hint'>(None)</span>";
    };

    row += "<td class='numbering'>&nbsp;</td>";
    row += "<td class='impexp'><input type='checkbox' /></td>";
    row += String.format('<td>{0}</td>', item.name);
    row += String.format('<td>{0}</td>', item.url);
    row += String.format('<td>{0}</td>', item.default_auth_method ? item.default_auth_method : "<span class='form_hint'>(None)</span>");
    row += String.format('<td>{0}</td>', service_name);
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.security.vault.connection.edit({0});'>Edit</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.security.vault.connection.delete_({0});'>Delete</a>", item.id));
    row += String.format("<td class='ignore item_id_{0}'>{0}</td>", item.id);
    row += String.format("<td class='ignore'>{0}</td>", item.timeout);
    row += String.format("<td class='ignore'>{0}</td>", item.allow_redirects);
    row += String.format("<td class='ignore'>{0}</td>", item.tls_verify);
    row += String.format("<td class='ignore'>{0}</td>", item.token);
    row += String.format("<td class='ignore'>{0}</td>", data.service_id);

    if(include_tr) {
        row += '</tr>';
    }

    return row;
}

$.fn.zato.security.vault.connection.delete_ = function(id) {
    $.fn.zato.data_table.delete_(id, 'td.item_id_',
        'Vault connection [{0}] deleted',
        'Are you sure you want to delete the Vault connection [{0}]?',
        true);
}
